app/src/main/java/http_server/RequestReader.java

```diff
@@ -0,0 +1,62 @@
+package http_server;
+
+import java.io.IOException;
+import java.util.HashMap;
+
+public class RequestReader {
+  private final SocketIO socketIO;
+  private String path;
+  private HashMap<String, String> headers;
+  private String body;
+
+  public RequestReader(SocketIO socketIO) {
+    this.socketIO = socketIO;
+  }
+
+  public String getPath() {
+    return path;
+  }
+
+  public String getBody() {
+    return body;
+  }
+
+  public void readRequest() throws IOException {
+    String rawRequest = getRequestHead();
+
+    if (!rawRequest.isBlank()) {
+      RequestParser parser = new RequestParser(rawRequest);
+      path = parser.path();
+      headers = parser.headers();
+
+      body = null;
+      if (headers.containsKey("Content-Length")) {
+        int bytes = Integer.parseInt(headers.get("Content-Length"));
+        body = getRequestBody(bytes);
+      }
+    }
+  }
+
+  private String getRequestHead() throws IOException {
+    StringBuilder request = new StringBuilder();
+    String line;
+
+    while ((line = socketIO.readLine()) != null) {
+      request.append(line).append("\r\n");
+
+      if (line.isEmpty()) {
+        break;
+      }
+    }
+
+    return request.toString();
+  }
```

Comment:
"Head" sounds really close to request Header but I know that was not your intention here. I don't quite know what another name would fit here but maybe something like "getStartOfRequest"? This is not a blocker, just a note.

Reply:
True! I'll change it in this next ticket since it's all dealing with the actual Headers.